### ISSUE TITLE
SBT fails to load resource defining implementation needed by `ServiceLoader`

### ISSUE THREAD

**Issue by bjhartin** [sbt/sbt#5726]
(Possible duplicate of #3498) A resource in a jar on the metabuild classpath is unavailable at runtime.

In my case it happens to be a `META-INF/services/XXX` file used when a third-party library invokes `ServiceLoader`.

## steps

To reproduce: 

- SBT versions 1.2.8 and 1.3.13 tested
- Clone the repository https://github.com/bjhartin/sbt-problem-reproduction
- Run the commands indicated in the `readme` of that repository
- Observe the `NoSuchElementException` which ultimately stems from the classloader's inability to locate the resource `META-INF/services/io.cucumber.core.gherkin.FeatureParser`
- Verify that this _should_ be reachable by executing the following sbt commands:
  - `sbt` (interactive shell) 
  - `reload plugins` (puts you in the metabuild)
  - `show Compile / fullClasspath`
    - Note that the `gherkin-messages` jar is present, which is where this resource resides

## problem

I think this may be due to the thread's context class loader being SBT's `MetaBuildClassloader` (observed via debugger).  It seems to be unable to load this resource.

## expectation

I would expect SBT to have runtime access to resources from its `Compile` dependencies.

## notes

I'm not using `sbt-cucumber` because it's stuck on an old version and upgrading it proved stubborn.

**Comment by eatkins** [comment]
Thanks for the report. Would you be able to produce a sample project that reproduces this issue?

**Comment by bjhartin** [comment]
Yes, I'll do that tomorrow.

**Comment by bjhartin** [comment]
@eatkins Here you go: https://github.com/bjhartin/sbt-problem-reproduction.

Instructions in the readme.

**Comment by bjhartin** [comment]
This problem also occurs if the plugin project is converted to just a library, depended on in `plugins.sbt` of the consuming project.

**Comment by bjhartin** [comment]
@eatkins I have a little more information now.  I've updated the  description and also simplified the repo with the reproduction.  It seems to have nothing to do with the plugin mechanism, see above.

**Comment by eatkins** [comment]
Thanks for the repro. I have an intuition about what is wrong and how to fix it but I can't say exactly when I'll get to looking at it.

**Comment by bjhartin** [comment]
@eatkins If you give me a pointer in the right direction I might be able to contribute, depending on how tricky it is.

**Comment by bjhartin** [comment]
Also, do you have a suggested workaround?  I tried hacking it by dropping the needed file in `project/src/main/resources`.  It appeared in the right place in `project/target/...`, but still could not be found by the classloader.

**Comment by eatkins** [comment]
I have no suggestions. I'm not convinced the problem is a classloading problem. I added logging to the plugin classloader in `findResource`, `findResources`, `getResource`, `getResources` and `getResourcesAsStream` and it never tried to load anything with gherkin in it. It did successfully find some resources in the io.cucumber package. The last thing I saw before it crashed was `found io/cucumber/htmlformatter/cucumber-react.css jar:file:/$COURSIER_LOCATION/v1/https/repo1.maven.org/maven2/io/cucumber/html-formatter/7.0.0/html-formatter-7.0.0.jar!/io/cucumber/htmlformatter/cucumber-react.css`

I'm not going to debug this further because I'd have to fork cucumber to get a better handle on things. My suggestion would be to open a bug with cucumber because they will have a lot more context than the sbt team on the expected behavior and could explain the assumptions that might help us fix this. I personally am not able to invest the time to solve this without external help.

**Comment by eatkins** [comment]
The relevant code is here: https://github.com/sbt/sbt/blob/31bea086aa9de6d39d760ce56a52acd61cf4d4cf/main/src/main/scala/sbt/internal/PluginManagement.scala#L70-L78
I also verified that two gherkin jars (cucumber-gherkin-6.2.2.jar and cucumber-gherkin-messages-6.2.2.jar) are added to the classpath. My best guess is that Cucumber is trying to load the service using a parent classloader because, as I said, I see no attempts to load gherkin resources in the plugin classloader.

**Comment by bjhartin** [comment]
@eatkins Thanks for looking, and I understand you can't spend tons of time on each issue.

For anyone else who encounters or tries to debug this issue, I think it is a classloader issue for two reasons:

- When I added a breakpoint at the service loader call in `FeatureParser`, it was using the `MetaBuildLoader`, whose list of classes included no dependency library classes (I don't know which CL loaded `FeatureParser`)
- The same issue was reported as #3498 ('meta-build resources not available to classes loaded as part of sbt plugins')

**Comment by eatkins** [comment]
Despite my previous comment, I am not saying that it isn't a classloading issue. But I'm not convinced that sbt is at fault. Even if it is, my intuition is that the issue can't be fixed in sbt. My guess is that cucumber is making invalid assumptions about the application classloader in the sbt context. It should be loading resources and classes from the plugin classloader, which has the `MetaBuildLoader` as a parent. The MetaBuildLoader cannot and should not be able to load the cucumber resources. I don't know why cucumber isn't using the plugin classloader which is why I suggested opening an issue with cucumber. The sbt team is very small and while we would like to address any and all issues, sometimes additional help is needed.

**Comment by bjhartin** [comment]
This is my own plugin - not cucumber's, so I think it can't be their issue.  I'm using a vanilla cucumber library and passing it args via `InputTask'

I think this is the same as #3498.

I can't post the plugin due to company IP rules, but it's dead simple - just uses `io.cucumber.core.runtime.Runtime.run` with the string args:

- This works fine if the plugin uses `runMain` and accepts a classname (the consuming project gives it an `App` wrapper around `io.cucumber.core.runtime.Runtime`)
- It doesn't work if the plugin uses `io.cucumber.core.runtime.Runtime` directly.  It will load the class fine but not the resources.

Thanks again.  I'm only belaboring this because I think it's a wider issue or a misunderstanding on my part.  I have a fairly acceptable workaround.

**Comment by eatkins** [comment]
Yeah, that makes sense. The runMain task will give you a classloader that is uncontaminated by the sbt classpath.

sbt tries to isolate plugins from each other. To do this, it gives each plugin a unique classloader that must have the sbt metabuild classloader as its parent. This is necessary for the plugin to implement the `AutoPlugin` api for sbt. My guess is that the cucumber runtime makes an assumption (possibly implicit) that it is being run by the application classloader. Depending on how cucumber internally manages classloaders, it could easily use one classloader for loading classes and another for loading resources. That is why I think the fix has to be made upstream in cucumber.

There isn't really a way for sbt to work without layered classloaders for plugins so I don't see how we can fix the issue on our end.

**Comment by steve-e** [comment]
I have run into a similar problem with ServiceLoader when trying to load a hadoop filesytem within an autoplugin.
